File: solutions/javascript/matching-brackets/matching-brackets.js
//
// Checks that all the brackets and braces in the string are matched correctly, and nested correctly
//
const matches = {
  "[": "]",
  "{": "}",
  "(": ")",
};

export const isPaired = (input) => {
  const closerNeeded = [];
  for (const c of input) {
    if (Object.values(matches).includes(c)) {
      // closing bracket was found, is it the next expected value on stack?
      if (closerNeeded.pop() != c) return false;
    } else if (Object.keys(matches).includes(c)) {
      // opening bracket was found, add matching closing value to the stack
      closerNeeded.push(matches[c]);
    }
  }
  return closerNeeded.length === 0;
};

File: solutions/python/matching-brackets/matching_brackets.py
"""Matching Brackets"""

MATCHES = {"[": "]", "{": "}", "(": ")"}


def is_paired(input_string):
    """Checks that all the brackets and braces in the string are matched correctly,
    and nested correctly"""
    closer_needed = []

    for c in input_string:
        if c in MATCHES.values():
            # closing bracket was found, is it the next expected value on stack?
            if not closer_needed or closer_needed.pop() != c:
                return False
        elif c in MATCHES:
            # opening bracket was found, add matching closing value to the stack
            closer_needed.append(MATCHES[c])

    return not closer_needed
